### ISSUE TITLE
Very high upload time on AWS S3 buckets because of filename check

### ISSUE THREAD

**Issue by kringkaste** [craftcms/cms#6253]
### Description

After migrating a customer from Craft 2 to Craft 3.4.24 we were faced with a serious problem on uploading assets to a AWS S3 bucket. The site is hosted behind AWS CF and when the customer uploaded an asset to a volume, the request was canceled by CF after the 60 seconds max origin respond time. So the customer could not upload assets directly in an entry anymore.

We saw, that the upload requests continued on the server and the asset was written to the database and the volume after 2-3 minutes even on big machines. So the asset upload was working in the background while CF didn't wanna wait :(

So we have looked into the process of the upload and identified the function which was this time consuming. It is:

https://github.com/craftcms/cms/blob/f33e994439417a3074940baa98e8865385b0b208/src/services/Assets.php#L791

With more than 130.000 assets in a folder of the S3 bucket, the function downloads a full list of all objects including all meta data as array, iterates over it, creating a new array with the same amount, testing it against another array of 130.000 database rows only to find a file with the same name :)

We are not sure, why this is build like this, but we had to do an emergency patch to make our customer happy again. So we replaced the full bucket download with:

```php
if ($volume->fileExists((string)$folder->path.'/'.$originalFilename)) {
    $existingFiles[mb_strtolower($originalFilename)] = true;
};
```

and let the rest as is. It's not a full refactoring, but this helps us to let a c5a.2xlarge performing the upload in 8 seconds than in 3 minutes :) We think this hole function should be refactored. We do not know enough of the background story, but we think the database query could also be simplified. You know the filename and could check this against the volume and the database without iterating over arrays, at least for filesystems supporting case-sensitive filenames.

And don't forget: codemonauts and our customers ❤️  Craft CMS 😄 

### Steps to reproduce

1. Have Craft 3.4.24 running behind AWS CF and using AWS S3 as volumes.
2. Have a lot of assets in a single folder of the volume (we have ~ 130.000 but we think the effect occurs much earlier).
3. Try to upload an additional asset and have some coffee.

### Additional info

- Craft version: 3.2.24
- PHP version: 7.4
- Plugins & versions: AWS S3 1.2.8

**Comment by andris-sevcenko** [comment]
OK. So I changed how this whole process works. Now, only files from the index that match the `<desired_filename><anything>.<extension>` pattern are fetched to improve performance on large asset counts in a single folder.

Further, Craft no longer fetches all the files in the folder from the volume to determine possible filenames. Unfortunately, Flysystem, which Craft uses, does not support filtering by filename as well as directory, so an approach described above could not be determined.

We went with changing the timestamp to `Y-m-d-His` pattern and just checking for each attempted filename in sequence This could result in more requests, but the payloads will be significantly smaller as well as using the new timestamp format should drastically reduce the chances of a conflicting filename.
